### ISSUE TITLE
Request or clarification: Multiple instances of the same engine

### ISSUE THREAD

**Issue by virtadpt** [searx/searx#1762]
I have kind of an unusual search case: I have two YaCy instances running on my server (one running as a 'regular' spider/search engine in Carusoe mode, one which only indexes and searches my file library on the same server for personal use only).  I'd like to configure Searx to be able to use both of them, differentiated by shortcodes (i.e., 'ya1' and 'ya2', or something like that).  This would require two copies of the YaCy stanza in searx/searx/settings.yml.

When I tried this, Searx kept throwing errors and wouldn't work.  I don't have the exact errors handy right now because I'm at work but I can get them tonight and add them to this ticket.

I'm working around this by running two instances of Searx on this box (I really only use the search API functionality of Searx) on separate ports, which works but is inelegant.

Has anyone gotten this kind of configuration to work?

**Comment by MarcAbonce** [comment]
I don't have a YaCy instance, but this should be possible if you add both pf them in your settings.yml with a different `name`, `shortcut` and `base_url`. The line with the `engine` is the only one that should stay the same.

**Comment by virtadpt** [comment]
Okay, let's give this a try...
```
  - name : Montauk
    engine : yacy
    shortcut : ya
    base_url : 'http://localhost:8090'
    number_of_results : 30
    timeout : 60.0

  - name : Baudrilliard
    engine : yacy
    shortcut : yb
    base_url : 'http://localhost:9999'
    number_of_results : 30
    timeout : 60.0
```
Okay, it's showing up in the preferences tab...
Accessing the first YaCy instance (!ya) worked.  Accessing the second (!yb) failed, and printed the following exception to stderr (I was running Searx from a shell (as searx:searx, because I normally run it as a system service)):
```
Traceback (most recent call last):
  File "/opt/searx/searx/searx/search.py", line 160, in search_one_http_request_safe
    search_results = search_one_http_request(engine, query, request_params)
  File "/opt/searx/searx/searx/search.py", line 97, in search_one_http_request
    return engine.response(response)
  File "/opt/searx/searx/searx/engines/yacy.py", line 65, in response
    raw_search_results = loads(resp.text)
  File "/usr/lib64/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
```

**Comment by virtadpt** [comment]
I'm guessing that nobody else is running into this problem?

**Comment by asciimoo** [comment]
@virtadpt using one engine module for multiple engines is perfectly fine. Good examples are the xpath and json engines.

**Comment by virtadpt** [comment]
@asciimoo I did some experiments yesterday, and a few of them do.  The Arch Wiki and Wikipedia engines worked.  I'll try to run some more tests at home tonight, but it seems that the Yacy engine doesn't.

**Comment by MarcAbonce** [comment]
Based on the error message you got, it looks like the result from one of the YaCy instances is not a valid JSON. Maybe if you print the `resp.text` just before the  [`raw_search_results = loads(resp.text)`](https://github.com/asciimoo/searx/blob/master/searx/engines/yacy.py#L64) line, you could get more information on what is YaCy actually returning. I think it could also help to print `params['url']` [at the end of the request function](https://github.com/asciimoo/searx/blob/master/searx/engines/yacy.py#L57) just to make sure that both YaCy urls are being queried separately.

**Comment by virtadpt** [comment]
I'll give that a try tonight after I get home.  Thanks!

**Comment by virtadpt** [comment]
Okay, I did some more tinkering and I think I got it a) working and b) stable.  One of the things I noticed in diffing the config files for my search bots was that the primary YaCy/Searx instance was using the bang-code `!ya` for YaCy.  When I added the second YaCy instance I gave it the bang-code `!yb` (one letter up from the default).  However, my bot was still configured to use the bang-code `!yz`, which would explain why it was never working right.  I'm going to hammer on it a bit more to make sure it's stable but I think that was it.
